### PR TITLE
endpoints for retrieving a donation of a user

### PR DESCRIPTION
**Pull request by nickhsine**
add three endpoints, including
- `/v1/users/:userID/periodic-donations/:id`, for retrieving a periodic donation according to its `id`
- `/v1/users/:userID/donations/prime/:id`, for retrieving a prime donation according to its `id`
- `/v1/users/:userID/donations/others/:id`, for retrieving a pay by other method donation according to its `id`

**TODO**
`/v1/users/:userID/donations/token/:id`, for retrieving a card token donation according to its `id`

### CODE REVIEW

File: controllers/donation.go [hunk 1]
```diff
@@ -39,14 +39,14 @@ type (
 	}
 
 	clientResp struct {
-		PeriodicID  null.Int          `json:"periodic_id"`
+		PeriodicID  uint              `json:"periodic_id,omitempty"`
```
Comment (babygoat) [inline controllers/donation.go]: Isn't this a required field?

File: controllers/donation.go [hunk 2]
```diff
@@ -39,14 +39,14 @@ type (
...
-		CardInfo    models.CardInfo   `json:"card_info"`
-		Cardholder  models.Cardholder `json:"cardholder"`
+		CardInfo    models.CardInfo   `json:"card_info,omitempty"`
+		Cardholder  models.Cardholder `json:"cardholder,omitempty"`
 		Amount      uint              `json:"amount"`
 		Currency    string            `json:"currency"`
-		Details     string            `json:"details"`
-		OrderNumber string            `json:"order_number"`
+		Details     string            `json:"details,omitempty"`
+		OrderNumber string            `json:"order_number,omitempty"`
```
Comment (babygoat) [inline controllers/donation.go]: see above

File: controllers/donation.go [hunk 3]
```diff
@@ -115,6 +115,69 @@ type (
 	}
 )
 
+func (cr *clientResp) BuildFromPeriodicDonationModel(pd models.PeriodicDonation) {
+	cr.CardInfo = models.CardInfo{
```
Comment (babygoat) [inline controllers/donation.go]: use `copier` package.

File: controllers/donation.go [hunk 4]
```diff
@@ -115,6 +115,69 @@ type (
...
+		Email:       pd.CardholderEmail,
+		Name:        pd.CardholderName,
+		NationalID:  pd.CardholderNationalID,
+		PhoneNumber: pd.CardholderPhoneNumber,
+		ZipCode:     pd.CardholderZipCode,
+	}
+	cr.Amount = pd.Amount
+	cr.Currency = pd.Currency
+	cr.PayMethod = defaultPeriodicPayMethod
+	cr.ID = pd.ID
```
Comment (babygoat) [inline controllers/donation.go]: use `copier` package?

File: controllers/donation.go [hunk 5]
```diff
@@ -115,6 +115,69 @@ type (
...
+		NationalID:  pd.CardholderNationalID,
+		PhoneNumber: pd.CardholderPhoneNumber,
+		ZipCode:     pd.CardholderZipCode,
+	}
+	cr.Amount = pd.Amount
+	cr.Currency = pd.Currency
+	cr.Details = pd.Details
+	cr.OrderNumber = pd.OrderNumber
+	cr.PayMethod = pd.PayMethod
+	cr.ID = pd.ID
```
Comment (babygoat) [inline controllers/donation.go]: see above

File: controllers/donation.go [hunk 6]
```diff
@@ -115,6 +115,69 @@ type (
...
+}
+
+func (cr *clientResp) BuildFromOtherMethodDonationModel(d models.PayByOtherMethodDonation) {
+	cr.Amount = d.Amount
```
Comment (babygoat) [inline controllers/donation.go]: see above.

File: routers/router.go [hunk 7]
```diff
@@ -87,11 +87,29 @@ func SetupRouter(cf *controllers.ControllerFactory) *gin.Engine {
...
+	v1Group.GET("/users/:userID/periodic-donations/:id", middlewares.CheckJWT(), middlewares.ValidateUserID(), ginResponseWrapper(func(c *gin.Context) (int, gin.H, error) {
+		return mc.GetADonationOfAUser(c, globals.PeriodicDonationType)
+	}))
 	v1Group.POST("/users/:userID/donations/:pay_method", middlewares.CheckJWT(), middlewares.ValidateUserID(), ginResponseWrapper(mc.CreateADonationOfAUser))
 	v1Group.PATCH("/users/:userID/donations/prime/:id", middlewares.CheckJWT(), middlewares.ValidateUserID(), ginResponseWrapper(func(c *gin.Context) (int, gin.H, error) {
 		return mc.PatchADonationOfAUser(c, globals.PrimeDonaitionType)
 	}))
 	// v1Group.GET("/users/:userID/donations", middlewares.CheckJWT(), middlewares.ValidateUserID(), ginResponseWrapper(mc.GetDonationsOfAUser))
+	// one-time donation including credit_card, line pay, apple pay, google pay and samsung pay
+	v1Group.GET("/users/:userID/donations/prime/:id", middlewares.CheckJWT(), middlewares.ValidateUserID(), ginResponseWrapper(func(c *gin.Context) (int, gin.H, error) {
```
Comment (babygoat) [inline routers/router.go]: The routing path seems to be ambiguous.  It seems to me that ad-hoc recordID for a user makes the semantic of url vague.  I suggest consolidating the listing endpoints into the same category (e.g, /v1/donations/:ID) and authorizing the request by the `id_token`. (user_id is a good candidate)

File: controllers/donation.go (hunk 1 above)
Comment (nickhsine) [reply controllers/donation.go]: it is a required field only if the `clientResp` is for card token donation.

File: controllers/donation.go (hunk 2 above)
Comment (nickhsine) [reply controllers/donation.go]: `Details` and `OrderNumber` is not specified in `PeriodicDonation`.

File: routers/router.go (hunk 7 above)
Comment (nickhsine) [reply routers/router.go]: `/v1/donations/:id` is ambiguous as well.
I will adopt `/v1/donations/prime/:id`.

Comment (nickhsine) [reply routers/router.go]: About the endpoints, I will change `GET` method in this PR. 
`PATCH` and `POST` will be fixed in the next PR.

Comment (nickhsine) [reply routers/router.go]: After discussion, 
the endpoints will be 
`/v1/periodic-donations/:recordID?user_id=userID`
`/v1/donations/prime/:recordID?user_id=userID`,
`/v1/donations/token/:recordID?user_id=userID`
`/v1/donations/others/:recordID?user_id=userID`.

**Comment by nickhsine** [comment]
@babygoat use copier and  router path change.

File: membership_user.sql [hunk 8]
```diff
@@ -244,6 +244,7 @@ CREATE TABLE `pay_by_prime_donations` (
   `card_info_country` varchar(30) DEFAULT NULL, 
   `card_info_country_code` varchar(10) DEFAULT NULL, 
   `card_info_expiry_date` varchar(6) DEFAULT NULL, 
+  `notes` varchar(100) DEFAULT NULL,
```
Comment (babygoat) [inline membership_user.sql]: Should we reserve the size for around 100 Chinese word or so?

Comment (nickhsine) [reply membership_user.sql]: sure, let me increase it to `256`?

File: membership_user.sql [hunk 9]
```diff
@@ -322,6 +323,7 @@ CREATE TABLE `periodic_donations` (
...
```
Comment (babygoat) [inline membership_user.sql]: see above.

File: membership_user.sql (hunk 8 above)
Comment (babygoat) [reply membership_user.sql]: Depends on the placeholder on the form of frontend

Comment (nickhsine) [reply membership_user.sql]: I prefer to design this field on the backend first, and notice frontend what size of this field we support.

Comment (babygoat) [reply membership_user.sql]: Then, reserve the size for 256 and put the 80-word constraint on frontend

Comment (nickhsine) [reply membership_user.sql]: by the way, I just check the default collation of utf8.
The default collation is `utf8_general_ci`, so `varchar(100)` is for 100 Chinese words tho.

**Comment by nickhsine** [comment]
@babygoat I will merge this PR if you think it is ok.

File: membership_user.sql (hunk 8 above)
Comment (babygoat) [reply membership_user.sql]: OK, I misunderstood the size which stands for character but not byte. BTW, collation is for the sorting order of the record which is encoded in the corresponding record. (utf-8 can have utf8_general_ci or utf8_encode_ci) I believe this has nothing to do with the size.

Comment (nickhsine) [reply membership_user.sql]: oh, I misunderstood collation as well.

**Pull request merged**